randomization/distribution_sample_generator.cc

```diff
@@ -0,0 +1,142 @@
+  }
+
+  auto prev_cdf = config.cdf_points(0).cdf();
+  if (prev_cdf < 0) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("The cdf value:'", prev_cdf,
+                     "' must not be negative in CDF:'", config.name(), "'."));
+  }
+
+  auto prev_value = config.cdf_points(0).value();
+  for (int i = 1; i < config.cdf_points_size(); i++) {
+    auto curr_value = config.cdf_points(i).value();
+    auto curr_cdf = config.cdf_points(i).cdf();
+    if (curr_value <= prev_value) {
+      return absl::InvalidArgumentError(absl::StrCat(
+          "The value:'", curr_value, "' must be greater than previous_value:'",
+          prev_value, "' at index '", i, "' in CDF:'", config.name(), "'."));
+    }
+    if (curr_cdf < prev_cdf) {
+      return absl::InvalidArgumentError(
+          absl::StrCat("The cdf value:'", curr_cdf,
+                       "' must be greater than previous cdf value:'", prev_cdf,
+                       "' at index '", i, "' in CDF:'", config.name(), "'."));
+    }
+    prev_value = curr_value;
+    prev_cdf = curr_cdf;
+  }
+
+  auto last_configured_cdf =
+      config.cdf_points(config.cdf_points_size() - 1).cdf();
+  if (last_configured_cdf != 1) {
+    return absl::InvalidArgumentError(absl::StrCat(
+        "The maximum value of cdf:'", last_configured_cdf, "' in CDF:'",
+        config.name(), "' must be exactly equal to 1."));
+  }
+  return absl::OkStatus();
+};
+
+absl::Status ValidateDistributionConfig(DistributionConfig& config) {
+  auto cdf_present = config.cdf_points_size() != 0;
+  auto pmf_present = config.pmf_points_size() != 0;
+
+  if (cdf_present == pmf_present) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("Exactly one of CDF and PMF must be provided for '",
+                     config.name(), "'."));
+  }
+
+  if (cdf_present) return ValidateCdfConfig(config);
+  if (pmf_present) return ValidatePmfConfig(config);
+
+  return absl::InvalidArgumentError(
+      absl::StrCat("Review CDF and PMF for '", config.name(), "'."));
+};
+
+std::unique_ptr<DistributionSampleGenerator> AllocateSampleGeneratorLibrary(
+    const DistributionConfig& config) {
+  std::unique_ptr<DistributionSampleGenerator> sample_gen_lib;
+  sample_gen_lib = std::make_unique<DistributionSampleGenerator>();
+  sample_gen_lib->Initialize(config);
+  return sample_gen_lib;
+};
+
+void DistributionSampleGenerator::InitializeWithCdf(
+    const DistributionConfig& config) {
+  DistributionConfig config_with_pmf;
+  config_with_pmf.set_name(config.name());
+
+  auto* pmf_point = config_with_pmf.add_pmf_points();
+  pmf_point->set_value(config.cdf_points(0).value());
+  pmf_point->set_pmf(config.cdf_points(0).cdf());
+
+  auto prev_cdf = config.cdf_points(0).cdf();
+  for (int i = 1; i < config.cdf_points_size(); i++) {
+    auto curr_cdf = config.cdf_points(i).cdf();
+    auto* pmf_point = config_with_pmf.add_pmf_points();
+    pmf_point->set_pmf(curr_cdf - prev_cdf);
```

Comment:
you never updated prev_cdf, and your unit tests never complained.

Reply:
The expectation here was that the USER will validate the config first before Initializing the sample generator. 
The test suite tested the ValidateDistributionConfig separately but did not test what happens when a wrong config was used to Initialze sample generator. I added these tests.

Lastly, yes, the current version produced wrong number of samples because the prev_cdf was not updated. Fixed this bug.